src/tools/rbd_mirror/GroupReplayer.cc

```diff
@@ -0,0 +1,964 @@
+    [this, on_finish](int r) {
+      if (r < 0) {
+	// Try start anyway.
+      }
+      start(on_finish, true, true);
+    });
+  stop(ctx, false, true);
+}
+
+template <typename I>
+void GroupReplayer<I>::flush() {
+  std::unique_lock locker{m_lock};
+  if (m_state != STATE_REPLAYING) {
+    return;
+  }
+
+  dout(10) << dendl;
+
+  for (auto &[_, image_replayer] : m_image_replayers) {
+    image_replayer->flush();
+  }
+}
+
+template <typename I>
+void GroupReplayer<I>::print_status(Formatter *f) {
+  dout(10) << dendl;
+
+  std::lock_guard l{m_lock};
+
+  f->open_object_section("group_replayer");
+  f->dump_string("name", m_group_spec);
+  auto state = m_state;
+  if (m_local_group_ctx.primary && state == STATE_REPLAYING) { // XXXMG
+    state = STATE_STOPPED;
+  }
+  f->dump_string("state", state_to_string(state));
+  f->open_array_section("image_replayers");
+  for (auto &[_, image_replayer] : m_image_replayers) {
+    image_replayer->print_status(f);
+  }
+  f->close_section(); // image_replayers
+  f->close_section(); // group_replayer
+}
+
+template <typename I>
+void GroupReplayer<I>::bootstrap_group() {
+  dout(10) << dendl;
+
+  std::unique_lock locker{m_lock};
+  if (m_peers.empty()) {
+    locker.unlock();
+
+    dout(5) << "no peer clusters" << dendl;
+    finish_start(-ENOENT, "no peer clusters");
+    return;
+  }
+
+  // TODO need to support multiple remote groups
+  ceph_assert(!m_peers.empty());
+  m_remote_group_peer = *m_peers.begin();
+
+  if (finish_start_if_interrupted(m_lock)) {
+    return;
+  }
+
+  ceph_assert(m_image_replayers.empty());
+
+  auto ctx = create_context_callback<
+      GroupReplayer,
+      &GroupReplayer<I>::handle_bootstrap_group>(this);
+  auto request = group_replayer::BootstrapRequest<I>::create(
+    m_threads, m_local_io_ctx, m_remote_group_peer.io_ctx, m_global_group_id,
+    m_local_mirror_uuid, m_instance_watcher, m_local_status_updater,
+    m_remote_group_peer.mirror_status_updater, m_cache_manager_handler,
+    m_pool_meta_cache, &m_resync_requested, &m_local_group_id,
+    &m_remote_group_id, &m_local_group_snaps, &m_local_group_ctx,
+    &m_image_replayers, &m_image_replayer_index, ctx);
+
+  request->get();
+  m_bootstrap_request = request;
+  locker.unlock();
+
+  set_mirror_group_status_update(
+    cls::rbd::MIRROR_GROUP_STATUS_STATE_STARTING_REPLAY, "bootstrapping");
+  request->send();
+}
+
+template <typename I>
+void GroupReplayer<I>::handle_bootstrap_group(int r) {
+  dout(10) << "r=" << r << dendl;
+  {
+    std::lock_guard locker{m_lock};
+    if (m_state == STATE_STOPPING || m_state == STATE_STOPPED) {
+      dout(10) << "stop prevailed" <<dendl;
+      return;
+    }
+    if (m_bootstrap_request != nullptr) {
+      m_bootstrap_request->put();
+      m_bootstrap_request = nullptr;
+    }
+    m_local_group_ctx.listener = &m_listener;
+    if (!m_local_group_ctx.name.empty()) {
+      m_local_group_name = m_local_group_ctx.name;
+    }
+  }
+
+  sync_group_names();
```

Comment:
This should be done only if the bootstrap succeeded.

Reply:
It is better to proceed to check for the names only if the bootstrap succeeded. The resync can be checked later in the code flow once it is determined that the bootstrap has got all the local and group information.